docs/design-docs/01-explicit-only-namespace-strategy.md

```diff
@@ -0,0 +1,101 @@
+# Explicit-only Namespace Strategy for `RootSync`s
+
+* Author(s): @tomasaschan
+* Approver: TBD
+* Status: provisional
+
+## Summary
+
+Config Sync currently has a concept of "implicit namespaces", where `RootSync`s can manage
+namespace-scoped resources without explicitly also defining the `Namespace` resource itself.
+In the current implementation, the `RootSync` will create the namespace and add ownership
+annotations in the same way as it would if the `Namespace` was explicitly managed.
+
+This can result in resource fights in the case where `RootSync`s implicitly use a namespace,
+if that namespace is also explicitly managed by another `RootSync`. This document outlines
+a mechanism to change this behavior to instead require explicit declarations of all namespaces
+_somewhere_, and allow resources to reconcile without causing resource fights.
+
+## Motivation
+
+Discussion about this was first sparked in [#191][191], where @tomasaschan set out to implement a
+solution for this. #207 outlines the use case without linking it to a suggested implementation.
+
+A simplified setup that exhibits the problems this design doc attempts to address, is the following:
+
+* A `RootSync` named `sync-a` declares `namespace-a` explicitly
+* A `RootSync` named `sync-x` declares resources in `namespace-a`, but not `namespace-a` itself
+* A `RootSync` named `sync-y` also declares resources in `namespace-a`, but not `namespace-a` itself
+* The reconciliation order of these syncs cannot be explicitly managed (e.g. because they are in turn
+  managed with a `RootSync`)
+
+If _either_ of `sync-x` or `sync-y` create `namespace-a` first, they will assume ownership of the
+namespace, and `sync-a` will not be allowed to create it (and, as a result, all its resources inside
+the namespace will also fail to be created).
+
+### Objective
+
+* The configuration described above should be supported without resource fights or reconciliation errors
+* Users who depend on implicit namespaces without also declaring them explicitly should need ot make
+  no changes to their configuration, and see the same behavior as today.
+
+## Design Overview
+
+The idea is to introduce an annotation that can be set on a `RootSync` which defines how to handle
+creation of namespaces: `configsync.gke.io/namespace-strategy`. The possible values are
+
+* `implicit` (the default): same behavior as today, including the failure modes triggering this proposal
+* `explicit`: namespaces must either exist in the cluster, or be declared in the configuration managed
+  by the `RootSync`; no implicit namespace creation happens.
+* (not set: same as `implicit`)
```

Comment:
Suggest making it a field to be more discoverable. It can be implemented as RootSync's `spec.override.namespaceStrategy`.

Reply:
Sure!

I suggested an annotation mainly based on [this comment thread](https://docs.google.com/document/d/1QK-vMQkcjmgKqaqI7fBBpejwr2eJsWu7lzO1q3PQqe4/edit?disco=AAAAu_xRu98) but I don't have any strong opinions either way. I'll update to use a spec field instead.